elements/lisk-chain/src/data_access/storage.ts

```diff
@@ -12,291 +12,537 @@
+		await this._db.clear({
+			gte: getFirstPrefix(DB_KEY_ACCOUNTS_ADDRESS),
+			lte: getLastPrefix(DB_KEY_ACCOUNTS_ADDRESS),
+		});
+		await this._db.clear({
+			gte: getFirstPrefix(DB_KEY_CHAIN_STATE),
+			lte: getLastPrefix(DB_KEY_CHAIN_STATE),
+		});
+		await this._db.clear({
+			gte: getFirstPrefix(DB_KEY_CONSENSUS_STATE),
+			lte: getLastPrefix(DB_KEY_CONSENSUS_STATE),
+		});
 	}
 
 	/*
 		Transactions
 	*/
+	public async getTransactionByID(id: string): Promise<TransactionJSON> {
+		const transaction = this._db.get<TransactionJSON>(
+			`${DB_KEY_TRANSACTIONS_ID}${id}`,
+		);
+
+		return transaction;
+	}
+
 	public async getTransactionsByIDs(
 		arrayOfTransactionIds: ReadonlyArray<string>,
 	): Promise<TransactionJSON[]> {
-		const transactions = await this._storage.entities.Transaction.get(
-			{
-				// eslint-disable-next-line camelcase
-				id_in: arrayOfTransactionIds,
-			},
-			{ limit: arrayOfTransactionIds.length },
-		);
+		const transactions = [];
+		for (const id of arrayOfTransactionIds) {
+			const transaction = await this.getTransactionByID(id);
+			transactions.push(transaction);
+		}
 
 		return transactions;
 	}
 
 	public async isTransactionPersisted(transactionId: string): Promise<boolean> {
-		const isPersisted = await this._storage.entities.Transaction.isPersisted({
-			id: transactionId,
-		});
+		return this._db.exists(`${DB_KEY_TRANSACTIONS_ID}${transactionId}`);
+	}
 
-		return isPersisted;
+	/*
+		Save Block
+	*/
+	public async saveBlock(
+		blockJSON: BlockJSON,
+		stateStore: StateStore,
+		removeFromTemp = false,
+	): Promise<void> {
+		const batch = this._db.batch();
+		const { transactions, ...header } = blockJSON;
+		batch.put(`${DB_KEY_BLOCKS_ID}${header.id}`, header);
+		batch.put(`${DB_KEY_BLOCKS_HEIGHT}${formatInt(header.height)}`, header.id);
+		if (transactions.length > 0) {
+			const ids = [];
+			for (const tx of transactions) {
+				ids.push(tx.id);
+				batch.put(`${DB_KEY_TRANSACTIONS_ID}${tx.id as string}`, tx);
+			}
+			batch.put(`${DB_KEY_TRANSACTIONS_BLOCK_ID}${header.id}`, ids);
+		}
+		if (removeFromTemp) {
+			batch.del(`${DB_KEY_TEMPBLOCKS_HEIGHT}${formatInt(blockJSON.height)}`);
+		}
+		stateStore.finalize(batch);
+		await batch.write();
+	}
+
+	public async deleteBlock(
+		blockJSON: BlockJSON,
+		stateStore: StateStore,
+		saveToTemp = false,
+	): Promise<void> {
+		const batch = this._db.batch();
+		const { transactions, ...header } = blockJSON;
+		batch.del(`${DB_KEY_BLOCKS_ID}${header.id}`);
+		batch.del(`${DB_KEY_BLOCKS_HEIGHT}${formatInt(header.height)}`);
+		if (transactions.length > 0) {
+			for (const tx of transactions) {
+				batch.del(`${DB_KEY_TRANSACTIONS_ID}${tx.id as string}`);
+			}
+			batch.del(`${DB_KEY_TRANSACTIONS_BLOCK_ID}${header.id}`);
+		}
+		if (saveToTemp) {
+			batch.put(
+				`${DB_KEY_TEMPBLOCKS_HEIGHT}${formatInt(blockJSON.height)}`,
+				blockJSON,
+			);
+		}
+		stateStore.finalize(batch);
+		await batch.write();
```

Comment:
If batch failed writing, are we not resetting the state store snapshot? Otherwise how the `getUpdated` identify changes to accounts which change after finalization?

Reply:
If batching is failed, then everything fails, and the state store instance will be thrown away